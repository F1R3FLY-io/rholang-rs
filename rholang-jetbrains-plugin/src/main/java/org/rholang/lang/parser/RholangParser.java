package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
