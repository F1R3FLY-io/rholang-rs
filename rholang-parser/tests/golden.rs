use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
