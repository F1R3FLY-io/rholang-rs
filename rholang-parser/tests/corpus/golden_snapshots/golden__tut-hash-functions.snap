---
source: rholang-parser/tests/golden.rs
expression: result
---
Good(
    [
        AnnProc {
            proc: New {
                decls: [
                    NameDecl {
                        id: Id {
                            name: "keccak256Hash",
                            pos: SourcePos {
                                line: 1,
                                col: 36,
                            },
                        },
                        uri: Some(
                            Uri(
                                "rho:crypto:keccak256Hash",
                            ),
                        ),
                    },
                    NameDecl {
                        id: Id {
                            name: "stdout",
                            pos: SourcePos {
                                line: 1,
                                col: 11,
                            },
                        },
                        uri: Some(
                            Uri(
                                "rho:io:stdout",
                            ),
                        ),
                    },
                    NameDecl {
                        id: Id {
                            name: "x",
                            pos: SourcePos {
                                line: 1,
                                col: 5,
                            },
                        },
                        uri: None,
                    },
                    NameDecl {
                        id: Id {
                            name: "y",
                            pos: SourcePos {
                                line: 1,
                                col: 8,
                            },
                        },
                        uri: None,
                    },
                ],
                proc: AnnProc {
                    proc: Par {
                        left: AnnProc {
                            proc: Par {
                                left: AnnProc {
                                    proc: Send {
                                        channel: AnnName {
                                            name: ProcVar(
                                                Id(
                                                    Id {
                                                        name: "x",
                                                        pos: SourcePos {
                                                            line: 2,
                                                            col: 5,
                                                        },
                                                    },
                                                ),
                                            ),
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 2,
                                                    col: 5,
                                                },
                                                end: SourcePos {
                                                    line: 2,
                                                    col: 6,
                                                },
                                            },
                                        },
                                        send_type: Single,
                                        inputs: [
                                            AnnProc {
                                                proc: Par {
                                                    left: AnnProc {
                                                        proc: Send {
                                                            channel: AnnName {
                                                                name: Quote(
                                                                    StringLiteral(
                                                                        "name",
                                                                    ),
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 2,
                                                                        col: 8,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 2,
                                                                        col: 15,
                                                                    },
                                                                },
                                                            },
                                                            send_type: Single,
                                                            inputs: [
                                                                AnnProc {
                                                                    proc: StringLiteral(
                                                                        "Joe",
                                                                    ),
                                                                    span: SourceSpan {
                                                                        start: SourcePos {
                                                                            line: 2,
                                                                            col: 17,
                                                                        },
                                                                        end: SourcePos {
                                                                            line: 2,
                                                                            col: 22,
                                                                        },
                                                                    },
                                                                },
                                                            ],
                                                        },
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 2,
                                                                col: 8,
                                                            },
                                                            end: SourcePos {
                                                                line: 2,
                                                                col: 23,
                                                            },
                                                        },
                                                    },
                                                    right: AnnProc {
                                                        proc: Send {
                                                            channel: AnnName {
                                                                name: Quote(
                                                                    StringLiteral(
                                                                        "age",
                                                                    ),
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 2,
                                                                        col: 26,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 2,
                                                                        col: 32,
                                                                    },
                                                                },
                                                            },
                                                            send_type: Single,
                                                            inputs: [
                                                                AnnProc {
                                                                    proc: LongLiteral(
                                                                        40,
                                                                    ),
                                                                    span: SourceSpan {
                                                                        start: SourcePos {
                                                                            line: 2,
                                                                            col: 34,
                                                                        },
                                                                        end: SourcePos {
                                                                            line: 2,
                                                                            col: 36,
                                                                        },
                                                                    },
                                                                },
                                                            ],
                                                        },
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 2,
                                                                col: 26,
                                                            },
                                                            end: SourcePos {
                                                                line: 2,
                                                                col: 37,
                                                            },
                                                        },
                                                    },
                                                },
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 2,
                                                        col: 8,
                                                    },
                                                    end: SourcePos {
                                                        line: 2,
                                                        col: 37,
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 2,
                                            col: 5,
                                        },
                                        end: SourcePos {
                                            line: 2,
                                            col: 38,
                                        },
                                    },
                                },
                                right: AnnProc {
                                    proc: ForComprehension {
                                        receipts: [
                                            [
                                                Linear {
                                                    lhs: Names {
                                                        names: [
                                                            AnnName {
                                                                name: Quote(
                                                                    ProcVar(
                                                                        Id(
                                                                            Id {
                                                                                name: "r",
                                                                                pos: SourcePos {
                                                                                    line: 3,
                                                                                    col: 15,
                                                                                },
                                                                            },
                                                                        ),
                                                                    ),
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 3,
                                                                        col: 14,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 3,
                                                                        col: 16,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                        remainder: None,
                                                    },
                                                    rhs: Simple {
                                                        name: AnnName {
                                                            name: ProcVar(
                                                                Id(
                                                                    Id {
                                                                        name: "x",
                                                                        pos: SourcePos {
                                                                            line: 3,
                                                                            col: 20,
                                                                        },
                                                                    },
                                                                ),
                                                            ),
                                                            span: SourceSpan {
                                                                start: SourcePos {
                                                                    line: 3,
                                                                    col: 20,
                                                                },
                                                                end: SourcePos {
                                                                    line: 3,
                                                                    col: 21,
                                                                },
                                                            },
                                                        },
                                                    },
                                                },
                                            ],
                                        ],
                                        proc: AnnProc {
                                            proc: Send {
                                                channel: AnnName {
                                                    name: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "keccak256Hash",
                                                                pos: SourcePos {
                                                                    line: 6,
                                                                    col: 13,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 6,
                                                            col: 13,
                                                        },
                                                        end: SourcePos {
                                                            line: 6,
                                                            col: 26,
                                                        },
                                                    },
                                                },
                                                send_type: Single,
                                                inputs: [
                                                    AnnProc {
                                                        proc: Method {
                                                            receiver: AnnProc {
                                                                proc: ProcVar(
                                                                    Id(
                                                                        Id {
                                                                            name: "r",
                                                                            pos: SourcePos {
                                                                                line: 6,
                                                                                col: 28,
                                                                            },
                                                                        },
                                                                    ),
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 6,
                                                                        col: 28,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 6,
                                                                        col: 29,
                                                                    },
                                                                },
                                                            },
                                                            name: Id {
                                                                name: "toByteArray",
                                                                pos: SourcePos {
                                                                    line: 6,
                                                                    col: 30,
                                                                },
                                                            },
                                                            args: [],
                                                        },
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 6,
                                                                col: 28,
                                                            },
                                                            end: SourcePos {
                                                                line: 6,
                                                                col: 43,
                                                            },
                                                        },
                                                    },
                                                    AnnProc {
                                                        proc: Eval {
                                                            name: AnnName {
                                                                name: ProcVar(
                                                                    Id(
                                                                        Id {
                                                                            name: "y",
                                                                            pos: SourcePos {
                                                                                line: 6,
                                                                                col: 46,
                                                                            },
                                                                        },
                                                                    ),
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 6,
                                                                        col: 46,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 6,
                                                                        col: 47,
                                                                    },
                                                                },
                                                            },
                                                        },
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 6,
                                                                col: 45,
                                                            },
                                                            end: SourcePos {
                                                                line: 6,
                                                                col: 47,
                                                            },
                                                        },
                                                    },
                                                ],
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 6,
                                                    col: 13,
                                                },
                                                end: SourcePos {
                                                    line: 6,
                                                    col: 48,
                                                },
                                            },
                                        },
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 3,
                                            col: 9,
                                        },
                                        end: SourcePos {
                                            line: 7,
                                            col: 10,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 2,
                                    col: 5,
                                },
                                end: SourcePos {
                                    line: 7,
                                    col: 10,
                                },
                            },
                        },
                        right: AnnProc {
                            proc: ForComprehension {
                                receipts: [
                                    [
                                        Linear {
                                            lhs: Names {
                                                names: [
                                                    AnnName {
                                                        name: Quote(
                                                            ProcVar(
                                                                Id(
                                                                    Id {
                                                                        name: "h",
                                                                        pos: SourcePos {
                                                                            line: 8,
                                                                            col: 15,
                                                                        },
                                                                    },
                                                                ),
                                                            ),
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 8,
                                                                col: 14,
                                                            },
                                                            end: SourcePos {
                                                                line: 8,
                                                                col: 16,
                                                            },
                                                        },
                                                    },
                                                ],
                                                remainder: None,
                                            },
                                            rhs: Simple {
                                                name: AnnName {
                                                    name: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "y",
                                                                pos: SourcePos {
                                                                    line: 8,
                                                                    col: 20,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 8,
                                                            col: 20,
                                                        },
                                                        end: SourcePos {
                                                            line: 8,
                                                            col: 21,
                                                        },
                                                    },
                                                },
                                            },
                                        },
                                    ],
                                ],
                                proc: AnnProc {
                                    proc: Send {
                                        channel: AnnName {
                                            name: ProcVar(
                                                Id(
                                                    Id {
                                                        name: "stdout",
                                                        pos: SourcePos {
                                                            line: 11,
                                                            col: 13,
                                                        },
                                                    },
                                                ),
                                            ),
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 11,
                                                    col: 13,
                                                },
                                                end: SourcePos {
                                                    line: 11,
                                                    col: 19,
                                                },
                                            },
                                        },
                                        send_type: Single,
                                        inputs: [
                                            AnnProc {
                                                proc: ProcVar(
                                                    Id(
                                                        Id {
                                                            name: "h",
                                                            pos: SourcePos {
                                                                line: 11,
                                                                col: 21,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 11,
                                                        col: 21,
                                                    },
                                                    end: SourcePos {
                                                        line: 11,
                                                        col: 22,
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 11,
                                            col: 13,
                                        },
                                        end: SourcePos {
                                            line: 11,
                                            col: 23,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 8,
                                    col: 9,
                                },
                                end: SourcePos {
                                    line: 12,
                                    col: 10,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 2,
                            col: 5,
                        },
                        end: SourcePos {
                            line: 12,
                            col: 10,
                        },
                    },
                },
            },
            span: SourceSpan {
                start: SourcePos {
                    line: 1,
                    col: 1,
                },
                end: SourcePos {
                    line: 13,
                    col: 2,
                },
            },
        },
    ],
)
